### PR TITLE
impl From<[T; N]> for Vec<T>

### LINKED ISSUE #67963: [Feature Request] implement From<array literal> for Vec
I would like the following code to compile:

```rust
fn f<S: Into<Vec<u8>> + Clone>(subscripts: &[S]) -> Vec<Vec<u8>> {
    subscripts.iter().cloned().map(|s| s.into()).collect()
}

fn main() {
    f(&[&b"hi"[..]]);  // this works
    f(&[b"hi"]); // error: `From<&[u8; 2]>` is not satisfied
}
```

The first call in main is hard to get right, since leaving out any part of the syntax gives confusing errors (without [..], gives `From<&&[u8; 2]>` is not satisfied, without & gives `doesn't have a size known at compile-time`).

The second call seems to be the same idea, but doesn't compile since From isn't implemented for array literals.

This can be implemented without trouble on nightly Rust except for the orphan rule: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=21632f9784745afa98663d5e6e471cbb

If someone is willing to mentor I'm happy to turn that playground into a PR :)

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/rust/issues/67963

### CODE REVIEW

File: src/liballoc/vec.rs
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
impl<T, const N: usize> From<[T; N]> for Vec<T>
where
    [T; N]: LengthAtMost32,
{
```

Comment (Centril) [reply src/liballoc/vec.rs]: Please also add a UI test in `src/test/ui/const-generics/array-impls/`.

File: src/liballoc/vec.rs
Comment (lzutao) [inline src/liballoc/vec.rs]:
```suggestion
    fn from(arr: [T; N]) -> Self {
```

**Comment by jyn514** [comment]
This is blocked on me getting stage1 to compile locally so I can mess with the test suite. It might be a couple days, I have a slow laptop and my fast computer runs alpine which doesn't work with proc-macros.

**Comment by jyn514** [comment]
Didn't take as long as I thought, only an hour :wink: This is ready for review.

**Comment by jyn514** [comment]
Does anyone know why this is giving an error?

```
2020-01-31T20:25:18.5690130Z    Compiling alloc v0.0.0 (/checkout/src/liballoc)
2020-01-31T20:25:21.2480774Z error[E0308]: mismatched types
2020-01-31T20:25:21.2481479Z     --> src/liballoc/vec.rs:2399:9
2020-01-31T20:25:21.2481892Z      |
2020-01-31T20:25:21.2482209Z 2398 |     fn from(arr: [T; N]) -> Self {
2020-01-31T20:25:21.2482609Z      |                             ---- expected `vec::Vec<T>` because of return type
2020-01-31T20:25:21.2482957Z 2399 |         <[T]>::into_vec(box arr)
2020-01-31T20:25:21.2483368Z      |         ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `vec::Vec`, found struct `std::vec::Vec`
2020-01-31T20:25:21.2483955Z      = note: expected struct `vec::Vec<T>`
2020-01-31T20:25:21.2484259Z                 found struct `std::vec::Vec<T>`
```

**Comment by jyn514** [comment]
Is there anything I need to change? This has been sitting around for a little while...

File: src/liballoc/vec.rs
Comment (Mark-Simulacrum) [inline src/liballoc/vec.rs]: Why is this gated on cfg(test)? AFAICT, it's not a lang item or so so that shouldn't matter? Could you tell us what errors you're getting without it?

**Comment by Mark-Simulacrum** [comment]
@withoutboats will need to take a look at this and decide whether to FCP merge it.

I think this is plausibly useful for a generic "takes `impl Into<Vec<T>>` type API, though mostly only because we lack `impl IntoIterator<Item=T>` on arrays. If we had that, it would be the "likely better" variant of this, I think, since APIs written to take iterators are much more flexible.

File: src/liballoc/vec.rs
Comment (jyn514) [reply src/liballoc/vec.rs]: Without this I was getting strange errors in stage2 in test mode about Vec not being the same as std::vec::Vec. I took this fix from [some nearby code](https://github.com/jyn514/rust/blob/e6ad27b9ace19f47fe2c8d3108ad03b1385ec105/src/liballoc/vec.rs#L2370),  if there's a better way I'd be happy to change it.

Full error: https://github.com/rust-lang/rust/pull/68692#issuecomment-580897843

**Comment by jyn514** [comment]
> APIs written to take iterators are much more flexible.

In this case I did need an owned buffer, not just an Iterator. `impl IntoIterator` for arrays would definitely be useful, it just wouldn't have solved this particular problem: https://github.com/rust-lang/rust/issues/67963

File: src/liballoc/vec.rs
Comment (Mark-Simulacrum) [reply src/liballoc/vec.rs]: Hm, interesting, okay. Seems suspicious but since we're doing it elsewhere also fine for now.

**Comment by Mark-Simulacrum** [comment]
IntoIterator would let you collect into a vector, which in theory should be equally performant if you're starting out with a vector or array etc. Thought that may not be true today.

**Comment by jyn514** [comment]
@withoutboats do you know when you will get a chance to look at this?

File: src/liballoc/vec.rs
Comment (LukasKalbertodt) [inline src/liballoc/vec.rs]: This will not land before `1.44` by now (beta cutoff for `1.42` is some time today). So this should be adjusted to 1.44 right?

File: src/liballoc/vec.rs
Comment (LukasKalbertodt) [inline src/liballoc/vec.rs]: Shouldn't this second version also work in the `cfg(not(test))` case?

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Looks fine to me (apart from the two inline comments) and seems like a useful addition.

File: src/liballoc/vec.rs
Comment (jyn514) [reply src/liballoc/vec.rs]: This does not work with `cfg(not(test))`.

```
error[E0425]: cannot find function `into_vec` in module `crate::slice`
    --> src/liballoc/vec.rs:2399:23
     |
2399 |         crate::slice::into_vec(box s)
     |                       ^^^^^^^^ not found in `crate::slice`
     |
help: possible candidate is found in another module, you can import it into scope
     |
59   | use crate::slice::hack::into_vec;
```

Comment (jyn514) [reply src/liballoc/vec.rs]: Possible explanation here: https://github.com/rust-lang/rust/blob/828305d10143172e0418d5d0a7111822149b4919/src/liballoc/vec.rs#L1851

Comment (LukasKalbertodt) [reply src/liballoc/vec.rs]: And [here is said `hack` module](https://github.com/rust-lang/rust/blob/828305d10143172e0418d5d0a7111822149b4919/src/liballoc/slice.rs#L134-L138). I'm still not fully getting why this is necessary, but let's accept it for now.

File: src/liballoc/vec.rs
Comment (LukasKalbertodt) [inline src/liballoc/vec.rs]: I'm not quite sure what's the procedure on this and what this tidy check tries to achieve, but it feels like this file could easily be split. There are already multiple sections. It would be easy, for example, to move the iterator stuff out of there (the last section, starting [here](https://github.com/rust-lang/rust/blob/3477e67a92878adae48b975915cb7a5c21026cd4/src/liballoc/vec.rs#L2486-L2488)). It's kind of unfortunate that you have to do it now, since this is unrelated to your PR, but I'd prefer that over the ignore tidy annotation.

But just to get a second opinion: @BurntSushi @withoutboats What do you think?

**Comment by LukasKalbertodt** [comment]
Apart from the tidy stuff, this looks good to me. But this is insta-stable and requires an FCP merge. Since I can't start those, reassigning:

r? @withoutboats

File: src/liballoc/vec.rs
Comment (Centril) [reply src/liballoc/vec.rs]: The purpose of the tidy check is to avoid gigantic files with poor structure. I think doing it before this PR or after is fine, but someone should sign up for the work.

**Comment by sfackler** [comment]
@rfcbot fcp merge

**Comment by TimDiekmann** [comment]
Is this really supposed to be insta-stable?

**Comment by sfackler** [comment]
All trait implementations are always stable.

**Comment by Centril** [comment]
Moving back to the technical reviewer, r? @LukasKalbertodt @bors r=LukasKalbertodt

File: src/liballoc/vec.rs
Comment (pickfire) [inline src/liballoc/vec.rs:2405]: Any docs for this?

Comment (jyn514) [reply src/liballoc/vec.rs:2405]: Does it need documentation? It looks pretty self-explanatory to me.

**Pull request merged**
